#163518687 Create social authentication for the application

espoirMur opened the pull request:
#### What does this PR do?

Enable social authentication for the project .

#### Description of Task to be completed?

This PR create the following endpoints : 
 - /api/v1/auth/google : to enable social authentication via google 
- /api/v1/auth/facebook : to enable social authentication via facebook
-  /api/v1/auth/twitter : to enable social authentication via twitter 
#### How should this be manually tested
1.  Update your .env file with social authentication credentials CLIENT_ID and CLIENT_SECRET for all providers : 
 - For google follow [this guide](https://developers.google.com/adwords/api/docs/guides/authentication) 
 - For twitter follow [this guide ](https://themepacific.com/how-to-generate-api-key-consumer-token-access-key-for-twitter-oauth/994/)
- For facebook  [this guide ](https://goldplugins.com/documentation/wp-social-pro-documentation/how-to-get-an-app-id-and-secret-key-from-facebook/)

```
GOOGLE_CLIENT_ID= '*******'
GOOGLE_CLIENT_SECRET= '*******'
FACEBOOK_CLIENT_ID =  '*******'
FACEBOOK_CLIENT_SECRET =  '*******'
TWITTER_CLIENT_ID = '********'
TWITTER_CLIENT_SECRET = '*******'
```
2. Run the project and open each link in the description in your browser and check if you can login with your social networks.
#### Any background context you want to provide?
None  
#### What are the relevant pivotal tracker stories?
#163518687 
#### Questions:
Guys, If you know how this feature could be improved let me know, especially on the testing for social authentication with Twitter let me know.

Musigwa committed:
[ft  #163518689] adding changes from develop

[Feature #163518689] fix rebase conflicts

[ft #[163518690]] create unittest for rating

[ft #[163518690] fix issue with page not found middleware in index.js

[ft #[163518690]] fix all test for rating endpoint

[ft #[163518690]  update  the documentation

[ft  #[163518690]] remove console.log

[ft #[163518690]] fix issue that cause the test to fail

[ft #[163518690]] adding a before and after for the whole test suite

[ft #[163518690]] adding an after for the whole suite

[ft #[163518690]] fix eslint errors

[story-type #[pivotal-tracker-id]] Description of task that has been implemented.

[ft  #[163518690]] create slug for migration

[ft #[163518690]] fixing failing test

[ft  #[163518690]] Fix all tests

[ft  #[163518690]] adding average rating when returning all ratings

[ft #[163518690]] adding jsdoc comment

[ft #[163518690]] adding status code from helpers

[ft #[163518690]] rebase develop branch

[ft #[163518690]] update the documentation

[ft #[163518690]] change JWT_KEY to SECRET

[ft #[163518690]] fix mirror issue of formatting

[ft #[163518690]]  changing article to slug

[ft  #[163518690]] update the documenttaion for passeport

[ft  #[163518690]] update swagger.json

[ft #[163518690)]] refractor code to use slug instead of id

[ft #[163518690]] refractor tests

[ft  #[163518690]] fix eslint errors

[ft  #[163518690]] update the documentation

[ft #[163518690 ]] add console.log to debug

[ft #[163518690]] fixing issue causing build failed on travis

[ft #[163518690]] remove only on rating test

[ft #[163518690]] fix issue on edit a rating for an article

[ft #[163518690]] remove for the test

[ft #[163518690]] fiing houndci problem

[ft #[163518690]] fix issue after rebasing develop

[ft #[163518690]] remove console.log statement

[ft #[163518690]] remove console.log statement

[ft  #163518689] adding changes from develop

[Feature #163518689] fix rebase conflicts

[ft #[163518690]] create unittest for rating

[ft #[163518690] fix issue with page not found middleware in index.js

[ft #[163518690]] fix all test for rating endpoint

[ft #[163518690]  update  the documentation

[ft  #[163518690]] remove console.log

[ft #[163518690]] fix issue that cause the test to fail

[ft #[163518690]] adding a before and after for the whole test suite

[ft #[163518690]] adding an after for the whole suite

[ft #[163518690]] fix eslint errors

[story-type #[pivotal-tracker-id]] Description of task that has been implemented.

[ft  #[163518690]] create slug for migration

[ft #[163518690]] fixing failing test

[ft  #[163518690]] Fix all tests

[ft  #[163518690]] adding average rating when returning all ratings

[ft #[163518690]] adding jsdoc comment

[ft #[163518690]] adding status code from helpers

[ft #[163518690]] rebase develop branch

[ft #[163518690]] update the documentation

[ft #[163518690]] change JWT_KEY to SECRET

[ft #[163518690]] fix mirror issue of formatting

[ft #[163518690]]  changing article to slug

[ft  #[163518690]] update the documenttaion for passeport

[ft  #[163518690]] update swagger.json

[ft #[163518690)]] refractor code to use slug instead of id

[ft  #[163518690]] update the documentation

[ft #[163518690 ]] add console.log to debug

[ft #[163518690]] fixing issue causing build failed on travis

[ft #[163518687]] adding social authentification fields to the database

[ft #[163518687]] creating endpoint for social auth

[ft #[163518687]] Adding social authentification with google

[ft #[163518687]] Adding social auth with facebook

[ft  #[163518687]] Adding twitter authentication

[ft #[16351868]] adding social authentication test

[ft #163518687] fixing test on social login with google

[ft #[163518687]] fix merge conflict in index.js

[ft #[163518687]]  testing social authentication failing

[ft #[163518687]] add more test for social authentication

[ft #[163518687]] temporary commit

[ft  #[163518687]] fix issues related to twitter

[ft #[163518687]] fix social authentication test exept for twitter

[ft  #[163518687]] refractor mocking code

[ft #[163518687]] update social authentication login callbacks

[ft #[163518687] update get all ratings function

abayo-luc commented on controllers/auth/socials/createUserFromSocial.js:
```diff
@@ -0,0 +1,37 @@
+  if (profile.emails) {
+    email = profile.emails[0].value;
+  } else {
+    email = `${profile.id}@ah.com`;
```
why this? will this email be an existing email that someone can use to reset or to confirm his account?

espoirMur replied:
I thought about this while working on this, but sometimes you may find users data from social authentication without password?
What Should we do in that case?

abayo-luc commented on controllers/auth/socials/createUserFromSocial.js:
```diff
@@ -0,0 +1,37 @@
+      },
+      defaults: {
+        username:
+          profile.username || `${profile.name.familyName}_${profile.name.givenName}`,
```
what if the username already exists in the database? what do you think about adding a random string or something like a timestamp to make the new username really unique?

espoirMur replied:
Let me apply the idea of timestamp to make the username unique...

abayo-luc commented on controllers/auth/socials/facebook.js:
```diff
@@ -0,0 +1,26 @@
+    {
+      clientID: process.env.FACEBOOK_CLIENT_ID,
+      clientSecret: process.env.FACEBOOK_CLIENT_SECRET,
+      callbackURL: "http://localhost:3000/api/v1/auth/facebook/callback",
```
make the URL base an environment variable so that this will work both in development as well as in production.

abayo-luc commented on controllers/auth/socials/google.js:
```diff
@@ -0,0 +1,26 @@
+    {
+      clientID: process.env.GOOGLE_CLIENT_ID,
+      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+      callbackURL: "http://localhost:3000/api/v1/auth/google/callback",
```
use env variable

abayo-luc commented on controllers/ratingsController.js:
```diff
@@ -53,25 +63,35 @@ class RatingController {
+            raw: true,
+            where: { articleId: article.id }
+          });
+          const averageRating = await Rating.findAll({
```
I think it seems like a duplicate. You are querying twice through a single table. What do you think finding a way of adding a method on Rating model that would calculate the average?

espoirMur replied:
Looks, like a good idea but still adding it to the model I will be querying the database twice.

abayo-luc commented on tests/integration/socialAuth/google.test.js:
```diff
@@ -0,0 +1,57 @@
+chai.use(chaiHttp);
+
+describe("mocking social authentication with google", () => {
+  // mocking api call
```
remove this comment!

YvesIraguha commented on controllers/auth/socials/facebook.js:
```diff
@@ -0,0 +1,26 @@
+      const user = await createUserFromSocial(profile);
+      return user
+        ? done(null, user)
+        : done(new Error("An error occurs , please try again later"));
```
@espoirMur Thank you for providing a generic error message, but change that grammatical error it is `An error occurred, please try again` not `An error occurs...`

YvesIraguha commented on controllers/auth/socials/google.js:
```diff
@@ -0,0 +1,26 @@
+      const user = await createUserFromSocial(profile);
+      return user
+        ? done(null, user)
+        : done(new Error("An error occurs , please try again later"));
```
Refer to the previous comment to correct this grammatical error.

YvesIraguha commented on controllers/auth/socials/twitter.js:
```diff
@@ -0,0 +1,26 @@
+      const user = await createUserFromSocial(profile);
+      return user
+        ? done(null, user)
+        : done(new Error("An error occurs , please try again later"));
```
Refer to the previous comment to correct this error message.

YvesIraguha commented on migrations/20190212153355-adding-social-auth-field.js:
```diff
@@ -0,0 +1,36 @@
+    /*
+      Add altering commands here.
+      Return a promise to correctly handle asynchronicity.
+    */
```
Remove these comments

YvesIraguha commented on migrations/20190212153355-adding-social-auth-field.js:
```diff
@@ -0,0 +1,36 @@
+    /*
+      Add reverting commands here.
+      Return a promise to correctly handle asynchronicity.
+    */
```
Remove these comments

YvesIraguha commented on tests/integration/article.test.js:
```diff
@@ -231,7 +231,7 @@ describe("# Articles endpoints", () => {
   });
 });
-describe("Share Articles endpoints", () => {
+describe.skip("Share Articles endpoints", () => {
```
This part can be removed, because the tests are being skipped, and we no longer need them in the back0-end, they will be implemented in the front-end.

YvesIraguha commented on tests/integration/comment.test.js:
```diff
@@ -1,49 +1,16 @@
-});
-
-describe("Comment on an article", () => {
+describe.skip("Comment on an article", () => {
```
This tests should not be skipped, in order to generate the branch coverage which is above a threshold level.

YvesIraguha commented on tests/integration/socialAuth/index.test.js:
```diff
@@ -0,0 +1,13 @@
+
+const { expect } = chai;
+
+it("check if social login callback is workings", async () => {
```
Correct that spelling error `... login callback is working` not `... is workings`

YvesIraguha commented on tests/integration/socialAuth/twitter.test.js:
```diff
@@ -0,0 +1,55 @@
+        );
+        done();
+      });
+  });
```
@espoirMur For the sake of consistency, we agreed on adding space after each test case. Therefore, you can add a space after each test case.

YvesIraguha reviewed (commented):
@espoirMur you did a great job of including social auth. We need to figure out how we can mock this Twitter API because it affects the ability of this PR to be merged. It won't reach the threshold level if these tests regarding Twitter API are skipped. I am open to help you conduct further research on how to mock Twitter API.

espoirMur committed:
[ft #[163518687]] repalce mock with stub for twitter authentication.

espoirMur committed:
[ft  #[163518687]] add test for error 500 when authentication failed

espoirMur committed:
[ft #[163518687]] remove social login

[ft #[163518687]] fixing hound CI errors

[ft #[163518687]] fixing hound CI errors

[ft #[163518687] refractor code and tests for social authentication

[ft  #[163518687] delete unused files for social authentication

espoirMur committed:
[ft #[163518687]] fixing hound ci problems

espoirMur committed:
[ft #[163518687]] fix issue with tests

espoirMur committed:
[ft #[163518687]] fix issue with tests

espoirMur committed:
[ft #[163518687] add lost migration file
